### PR TITLE
Components: Add gap between icon and button text

### PR DESCRIPTION
**Pull request by iamtakashi**
I would like to reopen the discussion in #16371 to see a consensus on how to fix this issue.

In #15387, the gap between the icon and the button text was added to the icon as a margin. As a result, the icon is no longer centre when there is no button text.

screen shot 2018-03-07 at 03 46 08

The lack of a tag around a button text makes this tricky to fix with just CSS.

@folletto suggested a number of ways to tackle this issue here: https://github.com/Automattic/wp-calypso/pull/16371#issuecomment-321242727

I'm leaning towards the idea of adding a `span` around a button text. What's your opinion?

cc @folletto @drw158

### CODE REVIEW

**Comment by folletto** [comment]
I think it's a good step. It's a bit verbose, but at least it provides clean CSS and a clean markup. If we can come up with a better idea later we can of course, but I'd start here.

_Note: the current PR changes just one example, so it's not representative._

**Comment by iamtakashi** [comment]
@folletto Thanks for your comment. 

Should we add the `<span>` in all the buttons regardless whether the button has an icon or not?

**Comment by folletto** [comment]
I think so, to make it consistent. 

The issue is more: how to make it in a reliable way? As you suggested in a previous chat: do we have to update ALL the buttons then? That's a lot of work. Can it be automated somehow?

**Comment by iamtakashi** [comment]
Not that I'm advocating but just an idea. How about adding the span to all the buttons that have an icon manually? I don't see a lot of buttons that have an icon so it's feasible. Although we could forget to add it in future, it's better than a space that we see in the devdoc.

**Comment by folletto** [comment]
If you feel it's feasible, I'm ok to start there. :)

**Comment by iamtakashi** [comment]
I've added the 4px gap to a gridicon inside a button only when it's not the last child of the button. This is more flexible approach as the tag wrapping button text can be anything. Also, it fixes the icon only buttons without doing anything extra.

Before:
screen shot 2018-03-08 at 18 50 57

After:
screen shot 2018-03-08 at 18 53 28

We still need to add a wrapping tag to a button text like what I've done in `example.jsx` to take the advantage fully, but I think this is a good first step as the most of exsiting button with an icon has already a custom solution.

What's the next step? Shall we merge this? Shall we change the exsiting buttons with this approach or leave them as they are?

**Comment by drw158** [comment]
The pseudo class seems like a good solution to me, and the most programmatic option, even though it could have unintended consequences.

File: client/my-sites/people/people-list-section-header/index.jsx [hunk 1]
```diff
@@ -55,7 +55,9 @@ class PeopleListSectionHeader extends Component {
 				{ siteLink && (
 					<Button compact href={ siteLink } className="people-list-section-header__add-button">
 						<Gridicon icon="user-add" />
-						{ translate( 'Invite', { context: 'Verb. Button to invite more users.' } ) }
+						<span className="people-list-section-header__button-text">
```
Comment (drw158) [inline client/my-sites/people/people-list-section-header/index.jsx]: Did you mean to leave this change in here? Should this be a separate PR?

File: client/components/button/docs/example.jsx [hunk 2]
```diff
@@ -59,7 +59,8 @@ class Buttons extends React.PureComponent {
 					<div className="docs__design-button-row">
 						<Button>Button</Button>
 						<Button>
-							<Gridicon icon="heart" /> Icon button
+							<Gridicon icon="heart" />
```
Comment (drw158) [inline client/components/button/docs/example.jsx]: I think we should remove the span for all of these examples, and let the developer decide if they need the extra span around the text.

File: client/my-sites/people/people-list-section-header/index.jsx (hunk 1 above)
Comment (iamtakashi) [reply client/my-sites/people/people-list-section-header/index.jsx]: It wasn't an accident, but I agree, it's better to be in a separate PR.

File: client/components/button/docs/example.jsx (hunk 2 above)
Comment (folletto) [reply client/components/button/docs/example.jsx]: You mean remove all, or show in the examples some with and some without? (I'd do the second)

Comment (iamtakashi) [reply client/components/button/docs/example.jsx]: Without the `span`, there won't be the gap between the icon and the text.

For me, this PR has two purposes:
1. To remove the unwanted gap in an icon only buttons.
2. More importantly, to establish what's the prefered way to give a gap between the icon and the text. (until we rework the component so that we can specify an icon as a prop)

My opinion on the second point is that we shouldn't have to add the gap every time to the icon, but instead, wrap the text with something like a span, and the component CSS takes care of that. If this was our consensus the devdoc should show that, shouldn't it?

Comment (folletto) [reply client/components/button/docs/example.jsx]: Ah yes... I have misunderstood, I was considering there were scenarios when that wasn't the case, but you're right. The example should show how it works, so always wrapping.

**Comment by iamtakashi** [comment]
@folletto, @drw158 Sorry for the ping, but can you review this as an interim solution?

**Comment by folletto** [comment]
As this fixes an issue, I think design wise it's good.

**Comment by iamtakashi** [comment]
@folletto raised a good question — do we need the class `button-text`?

We don't technically need it. I'm fine with either way.

@gwwar Can I ask you for a review and opinion too?

**Comment by drw158** [comment]
Leaving the class in there seems ok me. LGTM

**Comment by gwwar** [comment]
I suppose another question that comes to mind is should we allow icon buttons without labels in general? I often cannot interpret the meaning of what's intended.

**Comment by gwwar** [comment]
Code looks reasonable to me 👍 . Do folks know if we have compact buttons without text in prod? We'll want to verify that folks don't have custom CSS that would cause visual regressions

**Comment by iamtakashi** [comment]
> Do folks know if we have compact buttons without text in prod? 

I don't think we use it often, but we have two buttons in Media section.

screen shot 2018-03-15 at 21 59 12

**Comment by folletto** [comment]
> I suppose another question that comes to mind is should we allow icon buttons without labels in general? I often cannot interpret the meaning of what's intended.

You're correct: we should as much as possible avoid icon-only _anything_. Icons aren't clear at all.

However, there are some scenarios (which should be kept at minimum) when this is needed, either because it's a secondary action with a clear icon, there's no space at all and we're ok taking the understandability hit, or... I can't come up with any other scenario.

Note also that as component `Button` could be combined with a larger, more complex layout. As such, the button might end up being just an icon (borderless) because everything else explains what it does. This is a more likely and useful scenario.

**Comment by iamtakashi** [comment]
Thank you all!
